MEN-5322: implement the IoT Manager Device State management APIs

tranchitella opened the pull request:
Changelog: title

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

kjaskiewiczz commented on api/http/management_state.go:
```diff
@@ -0,0 +1,124 @@
+		return
+	}
+	integrationID := c.Param(paramIntegrationID)
+	_ = integrationID
```
what does that do?

tranchitella replied:
left-over, sorry

kjaskiewiczz commented on api/http/router.go:
```diff
@@ -47,10 +47,11 @@ const (
+
+	APIURLDevice                 = "/devices/:id"
+	APIURLDeviceState            = APIURLDevice + "/state"
+	APIURLDeviceStateIntegration = APIURLDevice + "/state/:integrationId"
```
should integrationId be a path or query param?

tranchitella replied:
According to @alfrunes's docs, it's a path param.

kjaskiewiczz commented on app/app.go:
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve the integration")
+	} else if integration == nil {
+		return nil, ErrIntegrationNotFound
```
I think it's not handled by the management handlers

kjaskiewiczz commented on app/app.go:
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
+	case model.AzureIoTHub:
+		return a.GetDeviceStateIoTHub(ctx, deviceID, integration)
+	default:
+		return nil, ErrUnknownIntegration
```
I think it's not handled by the management handlers

kjaskiewiczz commented on app/app.go:
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to retrieve the integration")
+	} else if integration == nil {
+		return nil, ErrIntegrationNotFound
```
I think it's not handled by the management handlers

kjaskiewiczz commented on app/app.go:
```diff
@@ -181,3 +190,46 @@ func (a *app) DeleteIOTHubDevice(ctx context.Context, deviceID string) error {
+	case model.AzureIoTHub:
+		return a.SetDeviceStateIoTHub(ctx, deviceID, integration, state)
+	default:
+		return nil, ErrUnknownIntegration
```
I think it's not handled by the management handlers

alfrunes commented on model/device.go:
```diff
@@ -0,0 +1,35 @@
+type DeviceIntegration struct {
+	// ID is the integration ID
+	ID string `json:"id" bson:"id"`
+	// ConnectionString is the device's primary connection string
+	// nolint:lll
+	PrimaryConnectionString *ConnectionString `json:"primary_connection_string" bson:"primary_connection_string"`
+	// Secondary ConnectionString is the device's secondary connection string
+	// nolint:lll
+	SecondaryConnectionString *ConnectionString `json:"secondary_connection_string" bson:"secondary_connection_string"`
```
Do we need to store device credentials in this service?

tranchitella replied:
No, we don't, but I thought you want.
Let me remove it :)

alfrunes commented on model/device.go:
```diff
@@ -0,0 +1,35 @@
+// DeviceIntegration stores the integration details for a specific device
+type DeviceIntegration struct {
+	// ID is the integration ID
+	ID string `json:"id" bson:"id"`
```
#9 is using `uuid.UUID` for integration IDs.

alfrunes commented on model/device.go:
```diff
@@ -0,0 +1,35 @@
+type Device struct {
+	// ID is the device ID
+	ID string `json:"id" bson:"_id"`
+	// Integrations contains the list of integrations for this device
+	Integrations []DeviceIntegration `json:"integrations" bson:"integrations"`
```
We will probably at some point need to be able to query devices that belong to a given integration, in that case it we can do this efficiently if we have a multi key index on a field holding the list of integration IDs.
If we need additional metadata stored within the device document, I would propose adding another embedded document which maps integration IDs to meta-data.
```suggestion
	// Integrations contains the list of integrations for this device
	IntegrationIDs []uuid.UUID `json:"?" bson:"integration_ids"`
	
	Integrations map[string]MetaData `json:"integrations,omitempty" bson:"integrations,omitempty"`
```

tranchitella replied:
We don't need the meta data, let's go with the IntegrationIDs only.

tranchitella commented:
@alfrunes @kjaskiewiczz thanks for your reviews, ready for a second round.

I addressed all your concerns and requests. There is an additional change: now both GetDeviceStateIntegration and SetDeviceStateIntegration will look up the device by (device ID, integration ID) before proceeding with the calls to IoT Hub. This ensures that the device exists in Mender before issuing API requests to IoT hub with arbitrary device IDs.

kjaskiewiczz reviewed (approved):
looks good

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -36,15 +37,21 @@ import (
 
 var (
-	ErrFailedToGetSettings = errors.New("Failed to get settings")
+	ErrFailedToGetDevice      = errors.New("Failed to get device")
+	ErrFailedToGetIntegration = errors.New("Failed to get integration")
+	ErrFailedToGetSettings    = errors.New("Failed to get settings")
```
Nitpick: Error messages should not be capitalized.

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -265,3 +269,38 @@ func (db *DataStoreMongo) CreateIntegration(
+
+func (db *DataStoreMongo) GetDeviceByIntegrationID(
+	ctx context.Context,
+	deviceID string,
```
Unused argument

alfrunes commented on store/mongo/datastore_mongo.go:
```diff
@@ -265,3 +269,38 @@ func (db *DataStoreMongo) CreateIntegration(
 }
+
+func (db *DataStoreMongo) GetDevice(ctx context.Context, deviceID string) (*model.Device, error) {
+	return db.GetDeviceByIntegrationID(ctx, deviceID, uuid.Nil)
```
This function is not doing what it is intended for because deviceID is unused.

tranchitella replied:
right, sorry, I fixed it!

alfrunes reviewed (commented):
I'm currently rebasing my work on top of yours and I found a significant error in the database interface. I can fix those in my upcoming PR.

tranchitella commented:
> I'm currently rebasing my work on top of yours and I found a significant error in the database interface. I can fix those in my upcoming PR.

I force-pushed, sorry for the trouble.
